config.go

```diff
@@ -148,39 +160,79 @@ func (config TGFConfig) String() string {
 	return string(bytes)
 }
 
-func (config *TGFConfig) getAwsSession() (*session.Session, error) {
-	return session.NewSessionWithOptions(session.Options{
-		Profile:                 config.tgf.AwsProfile,
-		SharedConfigState:       session.SharedConfigEnable,
-		AssumeRoleTokenProvider: stscreds.StdinTokenProvider,
-	})
+func (config *TGFConfig) getAwsSession(duration int64) (result *session.Session, err error) {
+	if cachedSession != nil {
+		return cachedSession, nil
+	}
+	options := session.Options{
+		Profile:           config.tgf.AwsProfile,
+		SharedConfigState: session.SharedConfigEnable,
+		AssumeRoleTokenProvider: func() (string, error) {
+			fmt.Fprintf(os.Stderr, "Assume Role MFA token code: ")
+			v, err := terminal.ReadPassword(int(os.Stdin.Fd()))
+			fmt.Fprintln(os.Stderr)
+			return string(v), err
+		},
+	}
+	if duration > 0 {
+		options.AssumeRoleDuration = time.Duration(duration) * time.Second
+	}
+	defer func() {
```

Comment:
Hmm, I don't think you need that defer? If that code was inline, it would lead to clearer code IMO. Only people familiar with go would understand this now

Reply:
Since there are not many exit point, I could change it. But, that's the pattern used to put a value in cache. I do expect that people doing golang understand golang 😜 .

Reply:
I don't understand the pattern. It's a two line switch to remove the defer

```
result, err = session.NewSessionWithOptions(options)
{defer content}
return result, err
```

Reply:
I know. My point is that when you want to assign the result in a cache variable, doing it in a defer ensure that you will always do it even if your function have multiple exit points. Since it is not the case here, I changed the code as you suggested.

Reply:
Oh alright, I get what you mean